File: implementation/client/testgenesis-client-node/src/agent.ts
import {createGrpcTransport} from "@connectrpc/connect-node";
import {Client, createClient, Transport} from "@connectrpc/connect";
import {create} from "@bufbuild/protobuf";
import {createWritableIterable} from "@connectrpc/connect/protocol";
import {AgentHub, AgentRegistrationSchema, ListenRequest, ListenRequestSchema, SummarySchema, TestResponse, TestResultSchema, TestState, TestStatusSchema, TranslationResponse, TranslationResultSchema, TranslationState, TranslationStatusSchema} from "./generated/index.js";
import {AnyProcessor, ProcessorType, TestSessionProcessor, TranslationSessionProcessor, TestSessionContext, TranslationSessionContext} from "./context.js";

export interface AgentConfig {
    hubUrl: string;
    displayName: string;
}

export class Agent {
    private readonly transport: Transport;
    private readonly agentClient: Client<typeof AgentHub>;
    private readonly processors = new Map<ProcessorType, AnyProcessor>();
    private isShuttingDown = false;

    constructor(private readonly config: AgentConfig) {
        this.transport = createGrpcTransport({
            baseUrl: config.hubUrl,
        });
        this.agentClient = createClient(AgentHub, this.transport);
    }

    public registerTestProcessor(processor: TestSessionProcessor) {
        this.processors.set("test", processor);
    }

    public registerTranslationProcessor(processor: TranslationSessionProcessor) {
        this.processors.set("translation", processor);
    }

    public async start() {
        console.log(`[Agent] Ready: ${this.config.displayName} (${Array.from(this.processors.keys()).join(", ")})`);

        while (!this.isShuttingDown) {
            try {
                await this.runLifecycle();
            } catch (err) {
                if (this.isShuttingDown) break;
                console.error("[Agent][Fatal] Disconnected. Retrying in 5s...", err);
                await new Promise(r => setTimeout(r, 5000));
            }
        }
    }

    public shutdown() {
        this.isShuttingDown = true;
    }

    private async runLifecycle() {
        // 1. Identification & Capability Registration
        const registration = await this.agentClient.register(create(AgentRegistrationSchema, {
            displayName: this.config.displayName,
            capabilities: Array.from(this.processors.values()).map(p => p.getCapability()),
        }));

        const clientId = registration.clientId;
        console.log(`[Agent] Connected as ${registration.clientId}`);

        // 2. Control Stream Dispatcher
        const requestIterable = createWritableIterable<ListenRequest>();
        const pendingSessions = new Map<string, ProcessorType>();
        const listenStream = this.agentClient.listen(requestIterable, {
            headers: {"x-client-id": clientId}
        });
        await requestIterable.write(create(ListenRequestSchema, {event: {case: "ready", value: {}}}));

        for await (const response of listenStream) {
            const event = response.event;

            if (event?.case === "sessionProposal") {
                const proposal = event.value;
                const type = proposal.details.case as ProcessorType;

                console.log(`[Agent][${type}] Handling Proposal: ${proposal.sessionId}`);

                if (this.processors.has(type)) {
                    pendingSessions.set(proposal.sessionId, type);
                    await requestIterable.write(create(ListenRequestSchema, {
                        event: {
                            case: "sessionAcceptance",
                            value: {sessionId: proposal.sessionId, accepted: true}
                        }
                    }));
                } else {
                    console.warn(`[Agent][${type}] Rejecting: No processor registered for this type.`);
                    await requestIterable.write(create(ListenRequestSchema, {
                        event: {
                            case: "sessionAcceptance",
                            value: {sessionId: proposal.sessionId, accepted: false}
                        }
                    }));
                }
            } else if (event?.case === "sessionReady") {
                const sessionId = event.value.sessionId;
                const type = pendingSessions.get(sessionId);

                if (type) {
                    pendingSessions.delete(sessionId);
                    this.handleSessionReady(sessionId, type).catch(err => {
                        console.error(`[Agent][${type}] Session Handling Error (${sessionId}):`, err);
                    });
                }
            }
        }
    }

    private async handleSessionReady(sessionId: string, type: ProcessorType) {
        if (type === "test") {
            const processor = this.processors.get("test") as TestSessionProcessor;
            const responseIterable = createWritableIterable<TestResponse>();
            const stream = this.agentClient.execute(responseIterable, {
                headers: {"x-session-id": sessionId},
            });

            console.log(`[Agent][Test] Active: ${sessionId}`);

            // Wait for Init without closing the stream (avoiding 'for await...break')
            const streamIterator = stream[Symbol.asyncIterator]();
            const { value: initMsg, done } = await streamIterator.next();

            if (done || !initMsg) throw new Error("Stream closed before TestInit was received.");

            const context = new TestSessionContext(initMsg, responseIterable);
            try {
                await processor.process(sessionId, context);
            } catch (err: any) {
                await context.sendResult(create(TestResultSchema, {
                    status: create(TestStatusSchema, {
                        state: TestState.FAILED,
                        message: `Internal Agent Error: ${err.message || String(err)}`
                    }),
                    summary: create(SummarySchema, {
                        metadata: {
                            exception: String(err),
                            stack: err.stack || "No stack trace available"
                        }
                    })
                }));
                throw err;
            } finally {
                responseIterable.close();
            }
        } else if (type === "translation") {
            const processor = this.processors.get("translation") as TranslationSessionProcessor;
            const responseIterable = createWritableIterable<TranslationResponse>();
            const stream = this.agentClient.translate(responseIterable, {
                headers: {"x-session-id": sessionId},
            });

            console.log(`[Agent][Translation] Active: ${sessionId}`);

            // Wait for Init without closing the stream (avoiding 'for await...break')
            const streamIterator = stream[Symbol.asyncIterator]();
            const { value: initMsg, done } = await streamIterator.next();

            if (done || !initMsg) throw new Error("Stream closed before TranslationInit was received.");

            const context = new TranslationSessionContext(initMsg, responseIterable);
            try {
                await processor.process(sessionId, context);
            } catch (err: any) {
                await context.sendResult(create(TranslationResultSchema, {
                    status: create(TranslationStatusSchema, {
                        state: TranslationState.FAILED,
                        message: `Internal Agent Error: ${err.message || String(err)}`
                    }),
                    summary: create(SummarySchema, {
                        metadata: {
                            exception: String(err),
                            stack: err.stack || "No stack trace available"
                        }
                    })
                }));
                throw err;
            } finally {
                responseIterable.close();
            }
        }
    }
}

File: implementation/client/testgenesis-client-java/src/main/java/me/hsgamer/testgenesis/client/Agent.java
package me.hsgamer.testgenesis.client;

import com.google.protobuf.Empty;
import com.google.protobuf.Struct;
import com.google.protobuf.Value;
import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import io.grpc.Metadata;
import io.grpc.stub.MetadataUtils;
import io.grpc.stub.StreamObserver;
import me.hsgamer.testgenesis.client.context.TestSessionContext;
import me.hsgamer.testgenesis.client.context.TranslationSessionContext;
import me.hsgamer.testgenesis.client.processor.TestSessionProcessor;
import me.hsgamer.testgenesis.client.processor.TranslationSessionProcessor;
import me.hsgamer.testgenesis.uap.v1.*;

import java.util.Map;
import java.util.concurrent.*;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * UAP Agent client for Java.
 * Multithreaded, pure Java implementation using gRPC-Java.
 */
public class Agent {
    private static final Logger logger = Logger.getLogger(Agent.class.getName());

    private final String hubUrl;
    private final String displayName;
    private final Map<String, Object> processors = new ConcurrentHashMap<>();
    private final Map<String, String> pendingSessions = new ConcurrentHashMap<>();
    private final ExecutorService sessionExecutor = Executors.newVirtualThreadPerTaskExecutor();

    private volatile boolean shuttingDown = false;
    private ManagedChannel channel;
    private StreamObserver<ListenRequest> controlRequestObserver;

    public Agent(String hubUrl, String displayName) {
        this.hubUrl = hubUrl;
        this.displayName = displayName;
    }

    public void registerTestProcessor(TestSessionProcessor processor) {
        processors.put("test", processor);
    }

    public void registerTranslationProcessor(TranslationSessionProcessor processor) {
        processors.put("translation", processor);
    }

    public void start() {
        logger.info("[Agent] Starting: " + displayName);

        while (!shuttingDown) {
            try {
                runLifecycle();
            } catch (Exception e) {
                if (shuttingDown) break;
                logger.log(Level.SEVERE, "[Agent][Fatal] Disconnected. Retrying in 5s...", e);
                try {
                    Thread.sleep(5000);
                } catch (InterruptedException ie) {
                    Thread.currentThread().interrupt();
                    break;
                }
            }
        }
    }

    public void shutdown() {
        shuttingDown = true;
        if (channel != null) {
            channel.shutdownNow();
        }
        sessionExecutor.shutdownNow();
    }

    private void runLifecycle() throws Exception {
        String host = hubUrl;
        int port = 80;
        if (hubUrl.startsWith("http://")) {
            host = hubUrl.substring(7);
        } else if (hubUrl.startsWith("https://")) {
            host = hubUrl.substring(8);
            port = 443;
        }

        String[] parts = host.split(":");
        if (parts.length > 1) {
            host = parts[0];
            port = Integer.parseInt(parts[1]);
        }

        channel = ManagedChannelBuilder.forAddress(host, port)
            .usePlaintext() // For production, should be configurable
            .build();

        AgentHubGrpc.AgentHubBlockingStub blockingStub = AgentHubGrpc.newBlockingStub(channel);
        AgentHubGrpc.AgentHubStub asyncStub = AgentHubGrpc.newStub(channel);

        // 1. Identification
        AgentRegistration.Builder registrationBuilder = AgentRegistration.newBuilder()
            .setDisplayName(displayName);

        for (Object p : processors.values()) {
            if (p instanceof TestSessionProcessor tp) {
                registrationBuilder.addCapabilities(tp.getCapability());
            } else if (p instanceof TranslationSessionProcessor tp) {
                registrationBuilder.addCapabilities(tp.getCapability());
            }
        }

        AgentRegistrationResponse registration = blockingStub.register(registrationBuilder.build());
        String clientId = registration.getClientId();
        logger.info("[Agent] Connected as " + clientId);

        // 2. Control Stream
        Metadata metadata = new Metadata();
        Metadata.Key<String> clientIdKey = Metadata.Key.of("x-client-id", Metadata.ASCII_STRING_MARSHALLER);
        metadata.put(clientIdKey, clientId);

        AgentHubGrpc.AgentHubStub metaAsyncStub = asyncStub.withInterceptors(MetadataUtils.newAttachHeadersInterceptor(metadata));

        final CountDownLatch streamLatch = new CountDownLatch(1);

        this.controlRequestObserver = metaAsyncStub.listen(new StreamObserver<ListenResponse>() {
            @Override
            public void onNext(ListenResponse response) {
                handleListenResponse(response, asyncStub);
            }

            @Override
            public void onError(Throwable t) {
                logger.log(Level.WARNING, "[Agent] Listen stream error", t);
                streamLatch.countDown();
            }

            @Override
            public void onCompleted() {
                logger.info("[Agent] Listen stream completed by server");
                streamLatch.countDown();
            }
        });

        controlRequestObserver.onNext(ListenRequest.newBuilder()
            .setReady(Empty.getDefaultInstance())
            .build());

        streamLatch.await();
    }

    private void handleListenResponse(ListenResponse response, AgentHubGrpc.AgentHubStub stub) {
        if (response.hasSessionProposal()) {
            SessionProposal proposal = response.getSessionProposal();
            String sessionId = proposal.getSessionId();
            String type = proposal.getDetailsCase().name().toLowerCase();

            logger.info("[Agent][" + type + "] Handling Proposal: " + sessionId);

            boolean accepted = processors.containsKey(type);

            if (accepted) {
                pendingSessions.put(sessionId, type);
            } else {
                logger.warning("[Agent][" + type + "] Rejecting: No processor registered.");
            }

            controlRequestObserver.onNext(ListenRequest.newBuilder()
                .setSessionAcceptance(SessionAcceptance.newBuilder()
                    .setSessionId(sessionId)
                    .setAccepted(accepted)
                    .build())
                .build());

        } else if (response.hasSessionReady()) {
            String sessionId = response.getSessionReady().getSessionId();
            String type = pendingSessions.remove(sessionId);
            if (type != null) {
                sessionExecutor.submit(() -> handleSessionReady(sessionId, type, stub));
            }
        }
    }

    private void handleSessionReady(String sessionId, String type, AgentHubGrpc.AgentHubStub asyncStub) {
        Metadata metadata = new Metadata();
        Metadata.Key<String> sessionIdKey = Metadata.Key.of("x-session-id", Metadata.ASCII_STRING_MARSHALLER);
        metadata.put(sessionIdKey, sessionId);
        AgentHubGrpc.AgentHubStub sessionStub = asyncStub.withInterceptors(MetadataUtils.newAttachHeadersInterceptor(metadata));

        if ("test".equals(type)) {
            TestSessionProcessor processor = (TestSessionProcessor) processors.get("test");
            handleTestSession(sessionId, processor, sessionStub);
        } else if ("translation".equals(type)) {
            TranslationSessionProcessor processor = (TranslationSessionProcessor) processors.get("translation");
            handleTranslationSession(sessionId, processor, sessionStub);
        }
    }

    private void handleTestSession(String sessionId, TestSessionProcessor processor, AgentHubGrpc.AgentHubStub stub) {
        BlockingQueue<TestInit> initQueue = new LinkedBlockingQueue<>(1);

        StreamObserver<TestResponse> responseObserver = stub.execute(new StreamObserver<TestInit>() {
            @Override
            public void onNext(TestInit value) {
                initQueue.offer(value);
            }

            @Override
            public void onError(Throwable t) {
                logger.log(Level.SEVERE, "[Agent][Test] Stream error (" + sessionId + ")", t);
            }

            @Override
            public void onCompleted() {
                logger.info("[Agent][Test] Stream completed (" + sessionId + ")");
            }
        });

        try {
            logger.info("[Agent][Test] Active: " + sessionId);
            TestInit init = initQueue.poll(30, TimeUnit.SECONDS);
            if (init == null) throw new TimeoutException("Timeout waiting for TestInit");

            TestSessionContext context = new TestSessionContext(init, responseObserver);
            try {
                processor.process(sessionId, context);
            } catch (Exception e) {
                context.sendResult(TestResult.newBuilder()
                    .setStatus(TestStatus.newBuilder()
                        .setState(TestState.TEST_STATE_FAILED)
                        .setMessage("Internal Agent Error: " + e.getMessage())
                        .build())
                    .setSummary(Summary.newBuilder()
                        .setMetadata(Struct.newBuilder()
                            .putFields("exception", Value.newBuilder().setStringValue(e.toString()).build())
                            .build())
                        .build())
                    .build());
                throw e;
            } finally {
                responseObserver.onCompleted();
            }
        } catch (Exception e) {
            logger.log(Level.SEVERE, "[Agent][Test] Session processing error (" + sessionId + ")", e);
        }
    }

    private void handleTranslationSession(String sessionId, TranslationSessionProcessor processor, AgentHubGrpc.AgentHubStub stub) {
        BlockingQueue<TranslationInit> initQueue = new LinkedBlockingQueue<>(1);

        StreamObserver<TranslationResponse> responseObserver = stub.translate(new StreamObserver<TranslationInit>() {
            @Override
            public void onNext(TranslationInit value) {
                initQueue.offer(value);
            }

            @Override
            public void onError(Throwable t) {
                logger.log(Level.SEVERE, "[Agent][Translation] Stream error (" + sessionId + ")", t);
            }

            @Override
            public void onCompleted() {
                logger.info("[Agent][Translation] Stream completed (" + sessionId + ")");
            }
        });

        try {
            logger.info("[Agent][Translation] Active: " + sessionId);
            TranslationInit init = initQueue.poll(30, TimeUnit.SECONDS);
            if (init == null) throw new TimeoutException("Timeout waiting for TranslationInit");

            TranslationSessionContext context = new TranslationSessionContext(init, responseObserver);
            try {
                processor.process(sessionId, context);
            } catch (Exception e) {
                context.sendResult(TranslationResult.newBuilder()
                    .setStatus(TranslationStatus.newBuilder()
                        .setState(TranslationState.TRANSLATION_STATE_FAILED)
                        .setMessage("Internal Agent Error: " + e.getMessage())
                        .build())
                    .build());
                throw e;
            } finally {
                responseObserver.onCompleted();
            }
        } catch (Exception e) {
            logger.log(Level.SEVERE, "[Agent][Translation] Session processing error (" + sessionId + ")", e);
        }
    }
}
